recorder/src/test/scala/weco/concepts/recorder/IndexSourceTest.scala

```diff
@@ -0,0 +1,186 @@
+
+    IndexSource[TestDoc](elasticHttpClient = testClient, indexName = "test")
+      .runWith(TestSink[TestDoc]())
+      .request(1)
+      .expectError()
+
+    testClient.requests should contain(
+      HttpRequest(
+        method = HttpMethods.POST,
+        uri = Uri("/test/_pit?keep_alive=1m")
+      )
+    )
+  }
+
+  it("returns a page of search results") {
+    implicit val actorSystem: ActorSystem = ActorSystem("test")
+    implicit val ec: ExecutionContext = actorSystem.dispatcher
+    val testDocs = (1 to 1000).map(id => TestDoc(id.toString))
+    val testClient =
+      new TestElasticHttpClient(
+        handlePitCreation orElse
+          handleSearch(testDocs)
+      )
+
+    val results = IndexSource[TestDoc](
+      elasticHttpClient = testClient,
+      indexName = "test",
+      pageSize = 100
+    )
+      .runWith(TestSink[TestDoc]())
+      .request(100)
+      .expectNextN(100)
+
+    results shouldBe testDocs.slice(0, 100)
+  }
+
+  it("returns subsequent pages of search results") {
+    implicit val actorSystem: ActorSystem = ActorSystem("test")
+    implicit val ec: ExecutionContext = actorSystem.dispatcher
+    val testDocs = (1 to 1000).map(id => TestDoc(id.toString))
+    val testClient = new TestElasticHttpClient(
+      handlePitCreation orElse
+        handleSearch(testDocs)
+    )
+
+    val results = IndexSource[TestDoc](
+      elasticHttpClient = testClient,
+      indexName = "test",
+      pageSize = 100
+    )
+      .runWith(TestSink[TestDoc]())
+      .request(1000)
+      .expectNextN(1000)
+
+    results shouldBe testDocs
+  }
+
+  it("reuses its point-in-time ID") {
+    implicit val actorSystem: ActorSystem = ActorSystem("test")
+    implicit val ec: ExecutionContext = actorSystem.dispatcher
+    val testDocs = (1 to 1000).map(id => TestDoc(id.toString))
+    val testClient = new TestElasticHttpClient(
+      handlePitCreation orElse
+        handleSearch(testDocs)
+    )
+
+    IndexSource[TestDoc](
+      elasticHttpClient = testClient,
+      indexName = "test",
+      pageSize = 100
+    )
+      .runWith(TestSink[TestDoc]())
+      .request(1000)
+      .expectNextN(1000)
+
+    val usedPitIds = Future
+      .sequence(testClient.requests.collect {
+        case HttpRequest(HttpMethods.POST, uri, _, entity, _)
+            if uri.path.endsWith("_search") =>
+          Unmarshal(entity)
+            .to[String]
+            .map(ujson.read(_))
+            .map(_.opt[String]("pit", "id"))
+      })
+      .map(_.flatten)
+
+    whenReady(usedPitIds) { ids =>
+      ids.toSet.size shouldBe 1
+    }
+  }
+
+  it("replaces the point-in-time ID if a search updates it") {
+    implicit val actorSystem: ActorSystem = ActorSystem("test")
+    implicit val ec: ExecutionContext = actorSystem.dispatcher
+    val testDocs = (1 to 200).map(id => TestDoc(id.toString))
+    val testClient = new TestElasticHttpClient(
+      handlePitCreation orElse
+        handleSearch(testDocs, _.reverse) orElse
```

Comment:
I think the only substantial change in behaviour between all the testClients is in handleSearch. (some don't delete the PIT, but I don't think there's any harm if the client can, and it just doesn't get called?)

Would it be better to have a Recorder Test client object that takes handleSearch and sandwiches it between the two PIT functions.

Reply:
True, will give this a bit of a refactor